### DISCUSSION TITLE
chap 8-2 질문 정리

### DISCUSSION THREAD

**Discussion by sihyung92** [JS-OS-study/hyokyung-and-students#29] [General]
chap 8 - memory management

**Comment by inseo24** [comment]
<h3>느낀 점 & 궁금한 점 </h3>

- memory management 방법에 대해서 열심히 배웠는데 마지막에 교수님이 사실 운영체제는 여기서 하는 일이 없단다~ 하고 뒷통수 맞은 기분을 느꼈다.
- 정말 생각해보니 여기서 주된 내용은 논리적인 주소를 물리적인 주소로 변환하는 작업인데 운영체제가 여기서 무슨 일을 한단다~ 하는 설명이 하나도 없었다.
- 허무했지만 다음 virtual memory로 빌드업을 쌓는 모습을 보고 다시 열심히 강의를 들어야겠다고 생각했다.

<hr />
<h3>문제</h3>

Q. 프로세스가 메모리에 접근할 때 운영체제는 아무 일도 하지 않는다. 이를 처리하는 것은 ____의 역할이다. 
왜 이렇게 작동할까요? 연결해서 MMU의 역할은 무엇일까요?

Q. 물리 메모리는 ___이라는 일정 크기의 블록으로 나뉩니다.(한글로 3글자, 영어로 5 words)

**Reply by 1chz** [reply]
- 하드웨어, 논리 주소와 물리 주소를 변환해주는 것
- 프레임

**Reply by pullingoff** [reply]
- 하드웨어/ 이유: 주소 변환 시마다 os가 개입해야한다면 cpu가 운영체제로 왔다갔다해야하니까 비효율적이기 때문이다. / MMU의 역할은 주소 변환을 그때그때 해주는 하드웨어임
- 파티션..??

**Reply by inseo24** [reply]
2번째 문제 답이 프레임입니다!

**Comment by 1chz** [comment]
# 느낀점

---

- 그래서 현대 컴퓨팅 시스템에서는 어떤 방식을 사용 하는 건가 싶었는데, 찾아보니 `Paging`과 `Paged Segmentation`을 혼용하여 사용하는 것 같다
- TLB는 현대 컴퓨팅 시스템에서 필연적인 요소라는 글을 많이 봤는데, 이번 기회에 TLB의 중요성을 크게 느낀 것 같다
- 컴퓨터는 처리 속도를 빠르게 하거나, 메모리를 절약하는 방향으로 발전해왔다. 이때 처리 속도와 메모리 사용량은 정비례한다
- 운영체제의 핵심 목적은 CPU와 메모리 자원을 효율적으로 사용하는 것이며, 이를 더 좋게 하기 위해 발전해왔다

# 문제

---

- 다단계 페이지 테이블을 사용하면 페이지 테이블을 위해 사용되는 메모리 공간의 소모를 줄일 수 있지만, 메모리 접근 시간이 늘어난다. 이를 어떻게 해결 할 수 있는가?

**Reply by inseo24** [reply]
TLB를 이용하는데 캐시처럼 동작한다. 많이 쓰이는 page entry를 가지고 있는 page table을 저장해두어 빠르게 찾을 수 있다.

**Reply by pullingoff** [reply]
주소 변환을 돕는 캐시 메모리인 TLB를 사용하면 된다. 페이징 테이블의 일부를 저장해둔다.

**Comment by pullingoff** [comment]
# 느낀 점
페이징 기법과 세그먼테이션 기법의 공통점과 차이점을 비교해 생각하니 더 이해하기 쉬웠다.

# 문제 
1. 역방향 페이지 테이블(Inverted Page Table)은 기존의 페이지 테이블의 어떤 단점을 보완하는가?
2. 페이지를 공유할 때 Shared code의 조건 2가지는?
3. 세그먼트 테이블과 페이지 테이블의 차이점은? (base, limit을 사용해 서술할 것)

**Reply by 1chz** [reply]
1. 페이지 번호를 보고 페이지 테이블에서 그만큼 떨어진 PTE에 가는 방법이 불가
2. read-only, 동일 논리 주소
3. 세그먼트 테이블은 가변분할방식이기 때문에 각 세그먼트의 크기를 나타내는 limit이 존재

**Reply by inseo24** [reply]
1. 페이지 테이블의 크기가 줄어든다?
2. read only, 동일 논리 주소
3. 세그먼트는 의미 단위로 잘라서 세그먼트 길이가 균일하지 않음 그래서 세그먼트 크기를 limit에 담고 있음, 페이징 기법은 페이지 크기가 균일해서 offset의 크기가 페이지 크기로 결정됨

**Comment by sihyung92** [comment]
# 느낀 점
- 지난 시간에 2단계 테이블 설명만 하길래 정리 안 하고 다단계를 따로 찾았는데, 바로 다음 수업에서 언급하니 머쓱했다.
- 가상 메모리 할당 개념에 대해선 얕게 알고 있었는데, [참고 링크](https://www.youtube.com/watch?v=5pEDL6c--_k&t=67s) 왠지 다음 시간 즈음에 할 거 같다. 두근두근

# 문제
- Paging 기법의 장단점, Segmentation architecture의 장단점에 대해서 서술해주세요~!

**Reply by 1chz** [reply]
- Paging 
  - 장점 : 외부 단편화 문제 해결 / 페이지 공유 가능 (메모리 공간의 효과적 활용) / 페이지 보호 가능
  - 단점 : 페이지 테이블 사용으로 가변분할방식 대비 주소 바인딩에서 오버헤드 발생 (=속도 저하) / 페이지 테이블을 메인 메모리에 저장함으로 공간 낭비

- Segmentation 
  - 장점: 내부 단편화 문제 해결, code, data, stack등을 의미별로 분류 저장하므로, 공유와 보호가 상대적으로 쉽다
  - 단점: 외부 단편화 문제 발생

**Reply by pullingoff** [reply]
페이징 기법
- 장점: 다단계 페이징 기법의 경우 TLB로 메모리 접근 시간을 줄일 수 있음
- 단점: 다단계 페이징 기법을 사용하지 않을 경우 페이지에 메모리가 없어도 모든 논리주소에 대응하는 페이지테이블 엔트리를 만들어야해서 테이블 공간이 낭비될 우려가 있음.

세그먼테이션 기법
- 장점: 의미 단위로 세그먼트를 나누기 때문에 공유, 보안에 있어 효과적
- 단점: 세그먼트의 크기가 균일하지 않아 hole(external fragmentation) 발생

**Reply by inseo24** [reply]
segmentation
장점 - 공유, 보안이 페이징보다 효과적
단점 - 외부 조각 발생

페이징 기법
장점 - 외부 조간 문제 해결, 페이지 공유 가능(메모리 공간 절약)
단점 - 속도 저하 문제, 오히려 공간 낭비가 커짐
